composer/models/ssd/utils.py

```diff
@@ -0,0 +1,510 @@
+
+        # Transform format to ltrb
+        l, t, r, b = bboxes_in[:, :, 0] - 0.5 * bboxes_in[:, :, 2], \
+                     bboxes_in[:, :, 1] - 0.5 * bboxes_in[:, :, 3], \
+                     bboxes_in[:, :, 0] + 0.5 * bboxes_in[:, :, 2], \
+                     bboxes_in[:, :, 1] + 0.5 * bboxes_in[:, :, 3]
+
+        bboxes_in[:, :, 0] = l
+        bboxes_in[:, :, 1] = t
+        bboxes_in[:, :, 2] = r
+        bboxes_in[:, :, 3] = b
+
+        return bboxes_in, F.softmax(scores_in, dim=-1)
+
+    def decode_batch(self, bboxes_in, scores_in, criteria=0.45, max_output=200, nms_valid_thresh=0.05):
+        bboxes, probs = self.scale_back_batch(bboxes_in, scores_in)
+
+        output = []
+        for bbox, prob in zip(bboxes.split(1, 0), probs.split(1, 0)):
+            bbox = bbox.squeeze(0)
+            prob = prob.squeeze(0)
+            output.append(self.decode_single(bbox, prob, criteria, max_output, nms_valid_thresh=nms_valid_thresh))
+        return output
+
+    # perform non-maximum suppression
+    def decode_single(self, bboxes_in, scores_in, criteria, max_output, max_num=200, nms_valid_thresh=0.05):
+        # Reference to https://github.com/amdegroot/ssd.pytorch
+
+        bboxes_out = []
+        scores_out = []
+        labels_out = []
+
+        for i, score in enumerate(scores_in.split(1, 1)):
+            if i == 0:
+                continue
+
+            score = score.squeeze(1)
+            mask = score > nms_valid_thresh
+
+            bboxes, score = bboxes_in[mask, :], score[mask]
+            if score.size(0) == 0:
+                continue
+
+            score_sorted, score_idx_sorted = score.sort(dim=0)
+
+            # select max_output indices
+            score_idx_sorted = score_idx_sorted[-max_num:]
+            candidates = []
+
+            while score_idx_sorted.numel() > 0:
+                idx = score_idx_sorted[-1].item()
+                bboxes_sorted = bboxes[score_idx_sorted, :]
+                bboxes_idx = bboxes[idx, :].unsqueeze(dim=0)
+                iou_sorted = calc_iou_tensor(bboxes_sorted, bboxes_idx).squeeze()
+                # we only need iou < criteria
+                score_idx_sorted = score_idx_sorted[iou_sorted < criteria]
+                candidates.append(idx)
+
+            bboxes_out.append(bboxes[candidates, :])
+            scores_out.append(score[candidates])
+            labels_out.extend([i] * len(candidates))
+
+        bboxes_out, labels_out, scores_out = torch.cat(bboxes_out, dim=0), \
+                                             torch.tensor(labels_out,
+                                                          dtype=torch.long), \
+                                             torch.cat(scores_out, dim=0)
+
+        _, max_ids = scores_out.sort(dim=0)
+        max_ids = max_ids[-max_output:]
+        return bboxes_out[max_ids, :], labels_out[max_ids], scores_out[max_ids]
+
+
+class DefaultBoxes(object):
```

Comment:
This is a pretty important part of SSD. We should document the box generation better if we're going to stick with this implementation.